Add bare bone test for migration to 154 to 155

mokagio opened the pull request:
See discussion at https://github.com/wordpress-mobile/WordPress-iOS/pull/24494#discussion_r2057152526

Problem: How do we know the migration from model version 154 to 155 works, in particular because the module for the entities changed in 155?

Proposed solution: Add unit tests to exercise the 154 -> 155 migration

We don't even have to merge this, might be enough to just see them green in CI and make a note.

kean committed:
Move remaining types to WordPressData

kean committed:
Use Current Module for Swift-only types

kean committed:
Fix ReaderTeamTopic compilation

kean committed:
Add missing class names

kean committed:
Remove unused SiteInfo

kean committed:
Add Bundle.wordPressData and update DataMigratorTests

kean committed:
Update CoreDataMigrationTests

kean committed:
Fix unit tests (AppEnvironment required)

mokagio committed:
Add bare bone test for migration to 154 to 155

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Add `BlogAuthor` create and read in 154-155 migration test - Crash!

mokagio committed:
Ensure the test store does not get reused

mokagio committed:
Run the Keystone and Data unit tests on iOS 18.4 in CI

See the silent failure
https://buildkite.com/automattic/wordpress-ios/builds/27496 due to the
pipe

mokagio committed:
Add a test action to the Keystone scheme

mokagio committed:
Add various missing WordPressData and WordPressKit imports to Keystone

See how the following build failed
https://buildkite.com/automattic/wordpress-ios/builds/27498/steps?sid=01966692-5a11-457c-8c19-e75c0b550aa7

This commit does not add all the necessary imports

mokagio committed:
Update module settings to work from WordPressData in ALL models

First, convert all entities to "Global Namespace" with

```
sed -i '' -E 's/representedClassName="(WordPress)?(\.)?([^"]+)"/representedClassName="\3"/g' \
  Sources/WordPressData/Resources/WordPress.xcdatamodeld/WordPress\ *.xcdatamodel/contents
```

Then, set the entities that need the "Current Product Module", using
https://github.com/wordpress-mobile/WordPress-iOS/pull/24494 as the
source for which to pick

```
sed -i '' -E 's/representedClassName="(ManagedAccountSettings|BlogAuthor|BloggingPrompt|BloggingPromptSettings|BloggingPromptSettingsReminderDays|BlogSettings|DiffAbstractValue|DiffContentValue|DiffTitleValue|ManagedDomain|ManagedPerson|Plan|PlanFeature|PlanGroup|PublicizeConnection|PublicizeService|ReaderCard|Revision|RevisionDiff|Role)"/representedClassName=".\1"/g' \
  Sources/WordPressData/Resources/WordPress.xcdatamodeld/WordPress\ *.xcdatamodel/contents
```

Notice that
`Sources/WordPressData/Resources/WordPress.xcdatamodeld/WordPress 155.xcdatamodel/contents`
did not change, which is a point in favor of showing this scripted
change being correct.

crazytonyli commented on Sources/WordPressData/Resources/WordPress.xcdatamodeld/WordPress 153.xcdatamodel/contents:
```diff
@@ -814,7 +814,7 @@
         <relationship name="blog" maxCount="1" deletionRule="Nullify" destinationEntity="Blog" inverseName="roles" inverseEntity="Blog" syncable="YES"/>
     </entity>
-    <entity name="SharingButton" representedClassName="WordPress.SharingButton" syncable="YES">
+    <entity name="SharingButton" representedClassName="SharingButton" syncable="YES">
```
There is no `.` in the `SharingButton` class name, and a few other entities. Is that expected?

mokagio commented:
Closing after @kean's input [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/24494#discussion_r2058248418)

> The first thing the app does is migrate from the production model (154). It never loads the represented entities in memory. It uses NSManagedObject and KVC for migration. It's impossible to retroactively change the models and there is also no need to. The app has representedClassName values in the previous models with class name that no longer even exist in the codebase – it's what the migration is for.

👋
